### ISSUE TITLE
Throwing non std::exception derived exception

### ISSUE THREAD

**Issue by DraTeots** [JeffersonLab/JANA2#3]
Hi!

When I stumbled upon this code:
```c++
std::shared_ptr<const JEvent> JEventSource_example2::GetEvent(void)
{
	// Throw exception if we have exhausted the source of events
	static size_t Nevents = 0; // by way of example, just count 1000000 events
	if( ++Nevents > 1000000 ) throw JEventSource::RETURN_STATUS::kNO_MORE_EVENTS;
```
I felt a bit strange that something throws ints in c++11 (even if it is enums). I thought that I clearly see reasons to throw it like this here. ```JEventSource::RETURN_STATUS::kNO_MORE_EVENTS``` is absolutely self explanatory and easy to use. The catch block is going to be Jana internal and one anyway would have something like:

``` C++
try {
   GetEvent()
}
catch(RETURN_STATUS) {
// ... why we not able to read further
}
catch(...) {
// the rest of errors
}
```
But still throwing something not std::exception derived looks strange for c++11 and above! I went to internet trying to find some reasoning (maybe it is OK and the problem is me =) ). After some research ([some notable QA](https://softwareengineering.stackexchange.com/questions/305641/should-one-derive-inherit-from-stdexception)) I came to a conclusion that since JANA is a general library, all thrown exceptions better to be derived from std::exception just to not to surprise users (even skipping other reasons). 

I would propose an approach which I like to use in some other libraries. The idea is to implement construction through static functions something like this:

```c++
class EventReadError: public std::runtime_error
{
      explicit EventReadError(const char* message, status):  _msg_(message), _status(status) {}
      explicit EventReadError(const std::string& message, status): msg_(message), _status(status){}
      virtual ~EventReadError() throw (){}
      virtual const char* what() const throw (){ return msg_.c_str();}

       // So we create such wrapper functions for users convenience
       static EventReadError NoMoreEvents() {
            return EventReadError(
                "No more events in event source",
                 RETURN_STATUS::kNO_MORE_EVENTS
       }

       // ... Other statuses. We probably don't need them all now.

protected:
    /** Error message.
     */
    std::string _msg;
    RETURN_STATUS _status;
};
```
All is to have readability and convenience on the users side:

```c++
std::shared_ptr<const JEvent> JEventSource_example2::GetEvent(void)
{
	// Throw exception if we have exhausted the source of events
        // This now became very self explanatory and convenient to use:
	if( ++Nevents > 1000000 ) throw EventReadError::NoMoreEvents()
```

If you agree on the approach I would implement and PR this particular exception and can hunt others.

**Comment by faustus123** [comment]
Looking over the proposed code it is not clear to me that the user would be aware that the EventReadError is derived from std:runtime_error. If they were, I would argue that this may be even more confusing since the std exception classes are intended to communicate error conditions. In this particular case (JEventSource has finished reading all events from its source) we use exceptions as basic flow control. One could argue that this should never be done and reserve exceptions only for errors. However, it is done in this case simply to save one "if" statement AND to make the code a little more concise. If we did not use exceptions here, we would need GetEvent to return return or set some flag (in addition to the JEvent) in order to indicate that no more events are available from the source. 

In short, I'm not quite seeing how the proposed change improves the user experience.

**Comment by DraTeots** [comment]
>  it is not clear to me that the user would be aware that the EventReadError is derived from std:runtime_error

Users shoudn't care about std:runtime_error. The only thing they should know (and care), that any exception in any C++ code nowadays inherit from std::exception.  

It is like washing hands and wearing gloves by surgeons. 

There are good and bad practices. Just imagine some side person opens the code to see how some "readout" part is made and see: 

1. Control flow is handled by exceptions ([what is wrong with that](https://web.archive.org/web/20140430044213/http://c2.com/cgi-bin/wiki?DontUseExceptionsForFlowControl))

2. Exceptions are not inherited from std::exceptions ([some reasoning](https://softwareengineering.stackexchange.com/questions/305641/should-one-derive-inherit-from-stdexception/305642#305642)) 

3. It gives warnings:
    ```
    cert-err09-cpp: Clang-Tidy: Throw expression should throw anonymous temporary values instead. 
    cert-err61-cpp: Clang-Tidy: Throwing an exception whose type 'JEventSource::RETURN_STATUS' is not derived from 'std::exception'. 
    ```


P.S. [This code](https://gitlab.com/jlab-eic/g4e/blob/master/src/JLeicDetectorConstruction.cc) works and works right too. At the same time I feel, it is not the most important question/feature at the moment =)

**Comment by faustus123** [comment]
I read through the articles you cited. For the first one, the article itself is just a brief paragraph at the top and most of the page is a debate from many people, some posing counter examples and reasons why a strict rule is inappropriate. There seem to be several arguments that using exceptions for flow control in situations similar to this is completely appropriate.

The Stack Overflow reference actually makes an argument NOT to throw something that inherits from std:exception in this case. Well, their argument is to always use std:exception so they can write a single catch block later. Here, since we use the exception as normal flow control, we don't want it to be caught in the same block that deals with errors. Keep in mind that JEventSource::GetEvent is called only from the JANA library which catches JEventSource::RETURN_STATUS types. The user has no possibility to ever catch this. On the contrary, suppose the user calls something from within GetEvent that throws an actual std::exception due to an error. That exception should propagate up to a catch block set up to handle errors. 

I hate to say it, but I think those first two citations convince me even more that this is appropriate.

I'm not 100% sure how to respond to item 3. It comes from the clang-tidy special option hicpp which checks for strict adherence to the Perforce hic++ standard. JANA2 does not currently subscribe to that standard. 

I think this boils down to a philosophical debate over whether exceptions should be restricted to true error conditions or  just exceptional conditions.  I seem them as a versatile tool and subscribe to the latter. However, I can appreciate that others may disagree and prefer to limit their use. I don't believe there is consensus in the field on this though so we might have to agree to disagree on this one.
